### PR TITLE
Poll and Track controllers were added.

### PR DESCRIPTION
**Pull request by DeccemberGirl**

### CODE REVIEW

File: Implementations/TrackRepository.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        public void AddTrack(Track track)
+        {
+            _DataContext.Tracks.Add(track);
+            _DataContext.SaveChanges();
+        }
+
+        public void ChangeTrack(Track track)
+        {
+            var trackToChange =  _DataContext.Tracks.SingleOrDefault(t => t.Name == track.Name);
```
Comment (DatAnton) [inline Implementations/TrackRepository.cs:25]: If we will have a few tracks with the same name(what potentially is real as we don't have any validation and unique attribute for this column) we will get an exception here.
To my mind we need find track by id. In this way we also can to change its name. But I am not sure is it ok to change its name.. it's business logic part. We can discuss it in telegram.

File: Managers/PollManager.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        public PollManager(IPollRepository pollRepository, IAccountRepository accountRepository)
+        {
+            _PollRepository = pollRepository;
+            _AccountRepository = accountRepository;
+        }
+
+        public ValidationResult CreatePoll(PollDTO pollDTO)
+        {
+            if (pollDTO == null)
```
Comment (DatAnton) [inline Managers/PollManager.cs:21]: Maybe we need more validations? For example checking not empty fields, etc?

File: Managers/PollManager.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            return _PollRepository.GetPollById(id);
+        }
+
+        private Pool GetPollFromDto(PollDTO pollDTO)
+        {
+            return new Pool
+            {
+                Name = pollDTO.Name,
+                Description = pollDTO.Description,
+                CreatorId = _AccountRepository.GetAccountByUsername(pollDTO.CreatorName).Id,
```
Comment (DatAnton) [inline Managers/PollManager.cs:43]: It's not ok to pass creatorName via request as you can create it like another people or pass incorrect creatorName. You should determine authorized user and set him as creator.

File: Managers/TrackManager.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        public TrackManager(ITrackRepository trackRepository, IPollToTrackRepository pollToTrackRepository)
+        {
+            _TrackRepository = trackRepository;
+            _PollToTrackRepository = pollToTrackRepository;
+        }
+
+        public ValidationResult CreateTrack(TrackDTO trackDTO)
+        {
+            if (trackDTO == null)
```
Comment (DatAnton) [inline Managers/TrackManager.cs:21]: More validation?

File: Managers/TrackManager.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                return new ValidationResult("The track entity was empty.");
+            }
+
+            var trackToCreate = GetTrackFromDto(trackDTO);
+            _TrackRepository.AddTrack(trackToCreate);
+
+            return ValidationResult.Success;
+        }
+
+        public ValidationResult ChangeTrack(TrackDTO trackDTO)
```
Comment (DatAnton) [inline Managers/TrackManager.cs:32]: here you can add parameter trackId what user want to update

File: Implementations/TrackRepository.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+            trackToChange.Link = track.Link;
```
Comment (DatAnton) [inline Implementations/TrackRepository.cs:27]: if new link will be empty you will clear it. So you need either validate it in manager or update field value only if new value is not epmty

File: Managers/PollManager.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                return new ValidationResult("The poll entity was empty.");
+            }
+
+            var pollToCreate = GetPollFromDto(pollDTO);
+            _PollRepository.AddPoll(pollToCreate);
+
+            return ValidationResult.Success;
+        }
+
+        public Pool GetPollById(int id)
```
Comment (DatAnton) [inline Managers/PollManager.cs:32]: I am not sure we need this method.. Maybe we can just call _PollRepository.GetPollById(id); from controller.. We can discuss it in telegram if you want.. or no discuss if you don't want=)

**Review by DatAnton — COMMENTED** [review]
As a result it looks not bad. Especially if you work with desktop and not web. Hope you will not be angry because of amount of my comments. As for me I hope it will help you to be more experienced programmer in future=)

File: Implementations/TrackRepository.cs (hunk 6 above)
Comment (DeccemberGirl) [reply Implementations/TrackRepository.cs:27]: Resolved

File: Managers/TrackManager.cs (hunk 5 above)
Comment (DeccemberGirl) [reply Managers/TrackManager.cs:32]: Resolved

File: Managers/TrackManager.cs (hunk 4 above)
Comment (DeccemberGirl) [reply Managers/TrackManager.cs:21]: Resolved

File: Managers/PollManager.cs (hunk 2 above)
Comment (DeccemberGirl) [reply Managers/PollManager.cs:21]: Resolved

File: Controllers/PollController.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        public IActionResult GetTracksIdByPollId(int pollId)
+        {
+            var poll = _PollManager.GetPollById(pollId);
+
+            if (poll == null)
+            {
+                return BadRequest(new { errorText = "There is no poll with the given id." });
+            }
+
+            var tracksId = poll.PoolToTracks.Select(pollToTrack => pollToTrack.TrackId);
```
[the remark this answers is not in the document]
Comment (DeccemberGirl) [reply Controllers/PollController.cs:30]: Resolved

File: Implementations/PollRepository.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+        public void AddPoll(Pool poll)
+        {
+            _DataContext.Pools.Add(poll);
+            _DataContext.SaveChanges();
+        }
+
+        public Pool GetPollById(int id)
+        {
+            return _DataContext.Pools.Where(p => p.Id == id).Include(p => p.PoolToTracks).SingleOrDefault();
```
[the remark this answers is not in the document]
Comment (DeccemberGirl) [reply Implementations/PollRepository.cs:26]: Resolved

[the remark this answers is not in the document]
Comment (DeccemberGirl) [reply Implementations/PollRepository.cs:26]: Resolved

File: Managers/PollManager.cs (hunk 3 above)
Comment (DeccemberGirl) [reply Managers/PollManager.cs:43]: Resolved

**Comment by DeccemberGirl** [comment]
> As a result it looks not bad. Especially if you work with desktop and not web. Hope you will not be angry because of amount of my comments. As for me I hope it will help you to be more experienced programmer in future=)

I managed to build pascal from the first time, of course i`ll succeed in web!)

File: Managers/PollManager.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            return ValidationResult.Success;
+        }
+
+        public IEnumerable<int> GetTracksIdByPollId(int pollId)
+        {
+            var poll = _PollRepository.GetPollById(pollId, true);
+
+            if (poll == null)
+            {
+                return new List<int>(1) { 0 };
```
Comment (DatAnton) [inline Managers/PollManager.cs:48]: why? return null and that's all.

**Review by DatAnton — APPROVED** [review]

**Review by megablochesk — APPROVED** [review]
It`s okey

**Pull request merged**